### ISSUE TITLE
EdgeSplitter 'uniform' sampling of negative edges is not uniform

### ISSUE THREAD

**Issue by huonw** [stellargraph/stellargraph#838]
### Describe the bug

The `EdgeSplitter` class says that it samples negative edges uniformly https://stellargraph.readthedocs.io/en/latest/api.html#stellargraph.data.EdgeSplitter

> Negative edges are sampled at random by uniformly (for ‘global’ method) selecting two nodes in the graph and then checking if these edges are connected or not. If not, the pair of nodes is considered a negative sample. Otherwise, it is discarded and the process repeats. Alternatively, negative edges are sampled (for ‘local’ method) using DFS search at a distance from the source node (selected uniformly at random from all nodes in the graph) sampled according to a given set of probabilities.

However, this doesn't seem to be the case: the resulting sample doesn't seem to have the expected properties one would expect. For instance, in the example below, it has no collisions of the source IDs of the edges (i.e. there are no nodes that have multiple negative edges leaving from it), which is theoretically exceedingly unlikely given the number of nodes and size of the sample in a true uniform sample.

### To Reproduce

1. Run the notebook [edge splitter.ipynb.zip](https://github.com/stellargraph/stellargraph/files/4183403/edge.splitter.ipynb.zip) (whole notebook markdown-ified below)

### Observed behavior

```python
import networkx as nx

import stellargraph as sg
import numpy as np
import pandas as pd
```

Create a graph with `1e5` nodes, and `5e4` "real" edges between them, leaving a lot of negative edges (`1e10 - 5e4` &approx; `1e10`) and thus not disturbing the statistical properties too much.

```python
N = 100000
g = nx.Graph()
g.add_nodes_from(range(N))
g.add_edges_from((i, i+1) for i in range(N//2))

es = sg.data.EdgeSplitter(g)

assert g.number_of_nodes() == N
assert g.number_of_edges() == N/2

# edge density (among all possible edges)
g.number_of_edges() / N**2
```

>     5e-06

```python
_, edges, labels = es.train_test_split(p=0.5, method="global")
```

>     Removed 1000 edges
>     ...
>     Sampled 24000 negative examples
>     ** Sampled 25000 positive and 25000 negative edges. **

```python
# x is the an array where each element is the number of items in the corresponding bin, this
# returns a summary on the distribution of those counts
def proportions_of_items_per_bin(x):
     return pd.DataFrame({"proportions": np.bincount(x)/len(x)})
```

Look at the number of repeats in the sources nodes of the sampled negative edges: there's no repeats.

```python
negative_edges = edges[labels == 0]

source_collisions = np.bincount(negative_edges[:, 0])

proportions_of_items_per_bin(source_collisions).style
```

<table id="T_86d7c0e8_4c54_11ea_abc1_acde48001122" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >proportions</th>    </tr></thead><tbody><tr><th id="T_86d7c0e8_4c54_11ea_abc1_acde48001122level0_row0" class="row_heading level0 row0" >0</th><td id="T_86d7c0e8_4c54_11ea_abc1_acde48001122row0_col0" class="data row0 col0" >0.75</td></tr><tr><th id="T_86d7c0e8_4c54_11ea_abc1_acde48001122level0_row1" class="row_heading level0 row1" >1</th><td id="T_86d7c0e8_4c54_11ea_abc1_acde48001122row1_col0" class="data row1 col0" >0.25</td></tr></tbody></table>

Now do 5000 truly uniform samples of the same number of items chosen from `N` elements (i.e. 5000 instances of truly uniform sampling of sources).

This results in a lot more repeats, and a distribution that doesn't look at all like the one above. For example, based on these proportions, out of the 25000 negative edges above, we'd expect to see ~600 pairs of negative edges that have the same source, not 0.

```python
# empirical distribution on the number of collisions
samples = 5000
sampled_sources = np.random.randint(0, N, size=(samples, len(negative_edges)))
sampled_source_collision_counts = np.apply_along_axis(np.bincount, 1, sampled_sources,  minlength=N)
proportions_of_items_per_bin(sampled_source_collision_counts.ravel()).style
```

<table id="T_8d29dd8c_4c54_11ea_b900_acde48001122" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >proportions</th>    </tr></thead><tbody><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row0" class="row_heading level0 row0" >0</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row0_col0" class="data row0 col0" >0.778793</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row1" class="row_heading level0 row1" >1</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row1_col0" class="data row1 col0" >0.194718</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row2" class="row_heading level0 row2" >2</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row2_col0" class="data row2 col0" >0.0243279</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row3" class="row_heading level0 row3" >3</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row3_col0" class="data row3 col0" >0.00202772</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row4" class="row_heading level0 row4" >4</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row4_col0" class="data row4 col0" >0.000127406</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row5" class="row_heading level0 row5" >5</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row5_col0" class="data row5 col0" >6.446e-06</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row6" class="row_heading level0 row6" >6</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row6_col0" class="data row6 col0" >2.62e-07</td></tr><tr><th id="T_8d29dd8c_4c54_11ea_b900_acde48001122level0_row7" class="row_heading level0 row7" >7</th><td id="T_8d29dd8c_4c54_11ea_b900_acde48001122row7_col0" class="data row7 col0" >8e-09</td></tr></tbody></table>

What's probability of getting at _most_ one collision? https://en.wikipedia.org/wiki/Birthday_problem#Cast_as_a_collision_problem

So small that it underflows a normal 64-bit float.

```python
n = len(negative_edges)
d = N
np.exp(-np.float128(n)*(n-1)/(2*d))
```

>     7.656510149776382209e-1358

### Expected behavior

If we're stating the edges are sampled "by uniformly [...] selecting two nodes in the graph", the `proportions_of_items_per_bin(source_collisions).style` should match more closely to the empirical distribution.

### Environment

Operating system: `Darwin-18.6.0-x86_64-i386-64bit`
Python version:
```
3.6.9 (default, Jul 10 2019, 12:25:55) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)]
```
Package versions:
<details>

```
absl-py==0.8.0
appdirs==1.4.3
appnope==0.1.0
astor==0.8.0
atomicwrites==1.3.0
attrs==19.3.0
backcall==0.1.0
black==19.10b0
bleach==3.1.0
boto==2.49.0
boto3==1.9.230
botocore==1.12.230
cachetools==4.0.0
certifi==2019.9.11
chardet==3.0.4
Click==7.0
coverage==4.5.4
coveralls==1.8.2
cycler==0.10.0
decorator==4.4.0
defusedxml==0.6.0
docopt==0.6.2
docutils==0.15.2
entrypoints==0.3
gast==0.2.2
gensim==3.8.0
google-auth==1.10.0
google-auth-oauthlib==0.4.1
google-pasta==0.1.7
grpcio==1.23.0
h5py==2.10.0
idna==2.8
importlib-metadata==0.23
ipykernel==5.1.3
ipython==7.9.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
isodate==0.6.0
jedi==0.15.1
Jinja2==2.10.3
jmespath==0.9.4
joblib==0.13.2
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==5.3.4
jupyter-console==6.0.0
jupyter-core==4.6.1
Keras==2.2.5
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.0
kiwisolver==1.1.0
llvmlite==0.30.0
Mako==1.1.0
Markdown==3.1.1
MarkupSafe==1.1.1
matplotlib==3.1.1
mistune==0.8.4
more-itertools==7.2.0
mypy==0.750
mypy-extensions==0.4.3
nbconvert==5.6.1
nbformat==4.4.0
networkx==2.3
notebook==6.0.2
numba==0.46.0
numpy==1.17.2
oauthlib==3.1.0
opt-einsum==3.1.0
packaging==19.2
pandas==0.25.1
pandocfilters==1.4.2
parso==0.5.1
pathspec==0.6.0
pdoc3==0.7.2
pexpect==4.7.0
pickleshare==0.7.5
pluggy==0.13.1
prometheus-client==0.7.1
prompt-toolkit==2.0.10
protobuf==3.9.1
ptyprocess==0.6.0
py==1.8.0
py-cpuinfo==5.0.0
py4j==0.10.7
pyasn1==0.4.8
pyasn1-modules==0.2.7
pydot==1.4.1
Pygments==2.4.2
pyparsing==2.4.2
pyrsistent==0.15.6
pyspark==2.4.4
pyspark-stubs==2.4.0.post6
pytest==5.3.1
pytest-benchmark==3.2.2
pytest-cov==2.8.1
pytest-repeat==0.8.0
python-dateutil==2.8.0
pytz==2019.2
PyYAML==5.1.2
pyzmq==18.1.1
qtconsole==4.6.0
rdflib==4.2.2
regex==2019.12.9
requests==2.22.0
requests-oauthlib==1.3.0
rsa==4.0
s3transfer==0.2.1
scikit-learn==0.21.3
scipy==1.3.1
Send2Trash==1.5.0
six==1.12.0
smart-open==1.8.4
-e git+git@github.com:stellargraph/stellargraph.git@9ea6930d2479a1e74dd0a093aac6dc8254e38dd2#egg=stellargraph
tensorboard==2.0.2
tensorflow==2.0.0
tensorflow-estimator==2.0.1
termcolor==1.1.0
terminado==0.8.3
testpath==0.4.4
toml==0.10.0
tornado==6.0.3
traitlets==4.3.3
typed-ast==1.4.0
typing-extensions==3.7.4.1
urllib3==1.25.3
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.15.6
widgetsnbextension==3.5.1
wrapt==1.11.2
zipp==0.6.0

```

</details>

### Additional context

N/A

**Comment by PantelisElinas** [comment]
Hi @huonw 

the difference is that negative sampling is not implemented as the bit of code in the end of your description. The code responsible for selecting negative edges is the following (as you have probably already seen)

```
 num_iter = int(np.ceil(num_edges_to_sample / (1.0 * len(start_nodes)))) + 1
 for _ in np.arange(0, num_iter):
        self._random.shuffle(start_nodes)
        self._random.shuffle(end_nodes)
        for u, v in zip(start_nodes, end_nodes):
```

It basically retrieves the list of nodes (twice unfortunately) and then it shuffles the order and considers each pair as the start and end nodes of edges. If num_iter is 1, then you would never get a node repeated as neither a start or end node so you won't get two edges starting from the same node. 

In the short term, we can just update the docstring to not say "uniformly" so as to not confuse people, if it does. The rest of the description is accurate. We can update the method as necessary later.

P.

**Comment by huonw** [comment]
I'm taking this, because it'll be quick.
